pkg/storegateway/indexheader/encoding/encoding.go

```diff
@@ -0,0 +1,265 @@
+// comparing the contents with the CRC32 checksum stored in the last four bytes.
+// CheckCrc32 consumes the contents of this Decbuf.
+func (d *Decbuf) CheckCrc32(castagnoliTable *crc32.Table) {
+	if d.r.Len() <= 4 {
+		d.E = ErrInvalidSize
+		return
+	}
+
+	hash := crc32.New(castagnoliTable)
+	bytesToRead := d.r.Len() - 4
+
+	for bytesToRead > 0 {
+		maxChunkSize := 1024 * 1024 // TODO: what is a sensible size to use here?
+		chunkSize := math.Min(bytesToRead, maxChunkSize)
+
+		// TODO: Could this be replaced with a Peak() + Skip() to avoid allocation?
+		// TODO: pull byte slices from a pool rather than creating a new one every time?
+		b, err := d.r.Read(chunkSize)
+		if err != nil {
+			d.E = errors.Wrap(err, "read contents for CRC32 calculation")
+			return
+		}
+
+		if n, err := hash.Write(b); err != nil {
+			d.E = errors.Wrap(err, "write bytes to CRC32 calculation")
+			return
+		} else if n != len(b) {
+			d.E = fmt.Errorf("CRC32 calculation only wrote %v bytes, expected to write %v bytes", n, len(b))
+			return
+		}
+
+		bytesToRead -= len(b)
+	}
+
+	actual := hash.Sum32()
+	expected := d.Be32()
+
+	if actual != expected {
+		d.E = ErrInvalidChecksum
+	}
+}
+
+// Skip advances the pointer of the underlying FileReader by the given number
+// of bytes. If E is non-nil, this method has no effect. Skip-ing beyond the
+// end of the underlying FileReader will set E to an error and not advance the
+// pointer of the FileReader.
+func (d *Decbuf) Skip(l int) {
+	if d.E != nil {
+		return
+	}
+
+	err := d.r.Skip(l)
+	if err != nil {
+		d.E = err
+	}
+}
+
+// ResetAt sets the pointer of the underlying FileReader to the absolute
+// offset and discards any buffered data. If E is non-nil, this method has
+// no effect. ResetAt-ing beyond the end of the underlying FileReader will set
+// E to an error and not advance the pointer of FileReader.
+func (d *Decbuf) ResetAt(off int) {
+	if d.E != nil {
+		return
+	}
+
+	err := d.r.ResetAt(off)
+	if err != nil {
+		d.E = err
+	}
+}
+
+// UvarintStr reads varint prefixed bytes into a string and consumes them. The string
+// returned allocates its own memory may be used after subsequent reads from the Decbuf.
+// If E is non-nil, this method returns an empty string.
+func (d *Decbuf) UvarintStr() string {
+	return string(d.UvarintBytes())
+}
+
+// UvarintBytes reads varint prefixed bytes into a byte slice consuming them but without
+// allocating. The bytes returned are no longer valid after subsequent reads from the Decbuf.
+// If E is non-nil, this method returns an empty byte slice.
+func (d *Decbuf) UvarintBytes() []byte {
+	l := d.Uvarint64()
+	if d.E != nil {
+		return []byte{}
+	}
+
+	b, err := d.r.Peek(int(l))
+	if err != nil {
+		d.E = err
+		return []byte{}
+	}
+
+	if len(b) != int(l) {
+		d.E = ErrInvalidSize
+		return []byte{}
+	}
+
+	if b == nil {
+		return []byte{}
+	}
+
+	err = d.r.Skip(len(b))
+	if err != nil {
+		d.E = err
+		return []byte{}
+	}
+
+	return b
```

Comment:
It's safe to return and use _until_ the next read of the underlying `FileReader`. After that, the bytes are invalidated. `FileReader.Skip()` (via `bufio.Reader.Discard()`) is guaranteed not to read from disk as long as `n` is less than the amount of data buffered. Because of the `Peek(n)` call above, we know that at least `n` bytes are buffered. 

I was wondering about this today and was in the process of writing a unit test to verify. I'll push that test along with other code review changes.

Reply:
> It's safe to return and use until the next read of the underlying FileReader

I would add a comment to this function to explain that.

Reply:
The docs for this method mention it but I'll make it more prominent.